src/plugins/external_tool_support/src/ExternalToolManager.cpp

```diff
@@ -193,28 +190,36 @@ QString ExternalToolManagerImpl::addTool(ExternalTool *tool) {
         foreach (const QString &dependency, toolDependencies) {
             dependencies.insertMulti(dependency, tool->getId());
         }
+    }
+}
 
-        if (dependenciesAreOk(tool->getId()) && !tool->isValid()) {
-            if (tool->isModule()) {
-                QString masterId = tool->getDependencies().first();
+QString ExternalToolManagerImpl::addToolToListsAndReturnToolPath(const ExternalTool &tool) {
+    QString toolPath;
+    SAFE_POINT(nullptr != etRegistry, "The external tool registry is nullptr", toolPath);
+
+    QStringList toolDependencies = tool.getDependencies();
+    if (!toolDependencies.isEmpty()) {
+        if (dependenciesAreOk(tool.getId()) && !tool.isValid()) {
+            if (tool.isModule()) {
+                QString masterId = tool.getDependencies().first();
                 ExternalTool *masterTool = etRegistry->getById(masterId);
-                SAFE_POINT(masterTool, QString("An external tool '%1' isn't found in the registry").arg(masterId), "");
+                SAFE_POINT(nullptr != masterTool, QString("An external tool '%1' isn't found in the registry").arg(masterId), toolPath);
 
                 toolPath = masterTool->getPath();
             } else {
-                toolPath = tool->getPath();
+                toolPath = tool.getPath();
             }
-            validateList << tool->getId();
+            validateList << tool.getId();
         }
     } else {
-        if (!tool->isValid()) {
-            validateList << tool->getId();
-            toolPath = tool->getPath();
+        if (!tool.isValid()) {
```

Comment:
what if tool.isValid == true ?

The logic of how all states are covered is not clear and may have missed states.
To make it simpler I would suggest is to extract the code that searches for the master tool path into a separate method and focus the code that all possible states (tool.isValid|dependenciesOk|...) are covered and none is missed.